components/tidb_query_datatype/src/codec/data_type/mod.rs

```diff
@@ -257,7 +257,41 @@ macro_rules! impl_evaluable_type {
     };
 }
 
-impl_evaluable_type! { Int }
+impl Evaluable for Int {
+    const EVAL_TYPE: EvalType = EvalType::Int;
+
+    #[inline]
+    fn borrow_scalar_value(v: &ScalarValue) -> Option<&Self> {
+        match v {
+            ScalarValue::Int(x) => x.as_ref(),
+            ScalarValue::Enum(x) => x
+                .as_ref()
+                .map(|x| unsafe { &*(x.value_ref() as *const u64 as *const i64) }),
+            _ => unimplemented!(),
+        }
+    }
+
+    #[inline]
+    fn borrow_scalar_value_ref(v: ScalarValueRef) -> Option<&Self> {
+        match v {
+            ScalarValueRef::Int(x) => x,
+            ScalarValueRef::Enum(x) => {
+                x.map(|x| unsafe { &*(x.value_ref() as *const u64 as *const i64) })
```

Comment:
This unsafe looks unsound.

Because x is `EnumRef` which implements `Copy`, the `x` inside the closure is a local variable inside of a function. So I think this closure returns a reference to a local variable.

Reply:
@skyzh

Reply:
To fix the problem, I think `EnumRef` may be defined like

```rust
pub struct EnumRef<'a> {
    name: &'a [u8],
    value: &'a u64,
}
```

Reply:
I did not even remember EnumRef implements Copy. I'll check soon.

Reply:
Agree. This is unsound. Will be fixed soon.

Reply:
What about changing it to `i64`? cc @dyzsr

Reply:
Agree